typing/env.mli

```diff
@@ -436,9 +436,34 @@ val print_path: (Format.formatter -> Path.t -> unit) ref
 
 (** Folds *)
 
+val fold_values:
+  (string -> Path.t -> value_description -> 'a -> 'a) ->
+  Longident.t option -> t -> 'a -> 'a
+val fold_types:
+  (string -> Path.t -> type_declaration -> 'a -> 'a) ->
+  Longident.t option -> t -> 'a -> 'a
 val fold_constructors:
   (constructor_description -> 'a -> 'a) ->
   Longident.t option -> t -> 'a -> 'a
+val fold_labels:
+  (label_description -> 'a -> 'a) ->
+  Longident.t option -> t -> 'a -> 'a
+
+(** Persistent structures are only traversed if they are already loaded. *)
+val fold_modules:
+  (string -> Path.t -> module_declaration -> 'a -> 'a) ->
+  Longident.t option -> t -> 'a -> 'a
```

Comment:
Looking at the implementation, I notice that `fold_modules` will silently skip functors. Is this intended? It should probably be documented. Is there a way to fold on module-like names that also include functors?

I have the vague impression that the current implementation `fold_modules` should be renamed into `fold_module_components`, and that `fold_modules` should actually fold on all module declarations, including functors. Which of the two functions would Merlin actually want to use?

Reply:
I think you're misreading the function: `fold_modules f lid_opt env` calls `f` on modules under the optional prefix `lid_opt`. When that prefix is unset, it's folding on modules directly accessible in the environment; when it's set, it's expecting the prefix to point to a module itself, and will fold on its submodules if it's a structure, or not do anything if it's a functor (because there is nothing to fold on).

Reply:
To clarify: `f` will also be called on modules which happen to be functors.

Merlin does use `fold_modules` as it is, I don't see what another implementation could do.

Reply:
Ah, I had misunderstood indeed. (Note sure why this function has such a complex implementation...) Thanks for the clarification, this is all good.